### PR TITLE
PIM-3548: Do not use the absolute file path of a media

### PR DESCRIPTION
**Pull request by solivier**
| Q                    | A
| -------------------- | ---
| Bug fix?             | Yes
| New feature?         | No
| BC breaks?           | Yes
| CI currently passes? | Yes
| Tests pass?          | Yes
| Scenarios pass?      | Yes
| Checkstyle issues?*  | No
| PMD issues?**        | No
| Changelog updated?   | Yes
| Fixed tickets        | PIM-3548
| DB schema updated?   | Yes
| Migration script?    | Yes
| Doc PR               | -

### CODE REVIEW

File: app/DoctrineMigrations/Version20150112174239.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+        $this->addSql('ALTER TABLE pim_catalog_product_media DROP file_path');
+    }
+
+    public function down(Schema $schema)
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE pim_catalog_product_media ADD file_path VARCHAR(500) DEFAULT NULL');
```
Comment (Nuscly) [inline app/DoctrineMigrations/Version20150112174239.php]: 500 is too small I think.

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php [hunk 2]
```diff
@@ -139,7 +139,11 @@ public function handle(ProductMediaInterface $media, $filenamePrefix)
      */
     public function duplicate(ProductMediaInterface $source, ProductMediaInterface $target, $filenamePrefix)
     {
-        $target->setFile(new File($source->getFilePath()));
+        if (null === $path = $this->getFilePath($source)) {
+            return;
```
Comment (filipsalpe) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: Hmm, we just silently do nothing if the source media is not valid? Maybe an exception?

File: app/DoctrineMigrations/Version20150112174239.php (hunk 1 above)
Comment (solivier) [inline app/DoctrineMigrations/Version20150112174239.php]: it's what we are currently using by default in 1.2

File: app/DoctrineMigrations/Version20150112174239.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Version20150112174239 extends AbstractMigration
+{
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE pim_catalog_product_media DROP file_path');
```
Comment (fitn) [inline app/DoctrineMigrations/Version20150112174239.php]: Should not be hardcoded. Please take a look at: https://github.com/webignition/ContainerAwareMigration
I already talk about this stuff some months ago so maybe it's to discuss with the team but it's not a big deal to do that.

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php (hunk 2 above)
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: Yep I agree with that. $this->getFilePath() should return an exception in some cases not only when you don't have READ rights on the file.

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php [hunk 4]
```diff
@@ -373,6 +373,10 @@ protected function delete(ProductMediaInterface $media)
      */
     protected function fileExists(ProductMediaInterface $media)
     {
+        if (null === $media->getFilename()) {
```
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: Should never appears. Please look at these issues: https://akeneo.atlassian.net/browse/PIM-3017 https://akeneo.atlassian.net/browse/PIM-3434

File: src/Pim/Bundle/TransformBundle/Normalizer/Structured/MediaNormalizer.php [hunk 5]
```diff
@@ -34,9 +45,11 @@ public function normalize($object, $format = null, array $context = [])
             ];
         }
 
+
+
```
Comment (fitn) [inline src/Pim/Bundle/TransformBundle/Normalizer/Structured/MediaNormalizer.php]: :D

File: upgrades/1.2-1.3/mongodb/migrate_media.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+$db = $client->$database;
+$productCollection = new MongoCollection($db, $productCollName);
+
+$products = $productCollection->find();
+
+echo sprintf("Removing filePath from %s media...\n", $products->count());
+
+foreach ($products as $product) {
+    if (array_key_exists('values', $product)) {
+        $n = count($product['values']);
```
Comment (fitn) [inline upgrades/1.2-1.3/mongodb/migrate_media.php]: $countValues instead of $n

File: src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php (hunk 4 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/MediaManager.php]: Thank you for pointing out this case, as discussed, it will be fix in another PR.

**Comment by nidup** [comment]
GTM
